src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp

```diff
@@ -171,20 +464,259 @@ bool ControllerScriptEngineLegacy::initialize() {
+            TransformScreenFrameFunction{tranformFunction, typed});
+}
+
+bool ControllerScriptEngineLegacy::bindSceneToScreen(
+        const LegacyControllerMapping::ScriptFileInfo& qmlFile,
+        const QString& screenIdentifier,
+        std::shared_ptr<ControllerRenderingEngine> pScreen) {
+    // Like for Javascript, if the script is invalid, it should be watched so the user can fix it
+    // without having to restart Mixxx. So, add it to the watcher before
+    // evaluating it.
+    watchFilePath(qmlFile.file.absoluteFilePath());
+
+    auto pScene = loadQMLFile(qmlFile, pScreen);
+    if (!pScene) {
+        pScreen->stop();
+        std::move(pScreen)->deleteLater();
+        return false;
+    }
+    const QMetaObject* metaObject = pScene->metaObject();
+
+    extractTransformFunction(metaObject, screenIdentifier);
 
+    connect(pScreen.get(),
+            &ControllerRenderingEngine::frameRendered,
+            this,
+            &ControllerScriptEngineLegacy::handleScreenFrame);
+    m_renderingScreens.insert(screenIdentifier, pScreen);
+    m_rootItems.insert(screenIdentifier, pScene);
+    // In case a rendering issue occurs, we need to shutdown the controller
+    connect(pScreen.get(),
+            &ControllerRenderingEngine::stopping,
+            this,
+            &ControllerScriptEngineLegacy::shutdown);
     return true;
 }
 
+void ControllerScriptEngineLegacy::handleScreenFrame(
+        const LegacyControllerMapping::ScreenInfo& screeninfo,
+        const QImage& frame,
+        const QDateTime& timestamp) {
+    VERIFY_OR_DEBUG_ASSERT(
+            m_transformScreenFrameFunctions.contains(screeninfo.identifier) ||
+            m_renderingScreens.contains(screeninfo.identifier)) {
+        qCWarning(m_logger) << "Unable to find transform function info for the given screen";
+        return;
+    };
+    VERIFY_OR_DEBUG_ASSERT(m_rootItems.contains(screeninfo.identifier)) {
+        qCWarning(m_logger) << "Unable to find a root item for the given screen";
+        return;
+    };
+
+    if (CmdlineArgs::Instance().getControllerPreviewScreens()) {
+        QImage screenDebug(frame);
+
+        switch (screeninfo.endian) {
+        case LegacyControllerMapping::ScreenInfo::ColorEndian::Big:
+            qFromBigEndian<ushort>(frame.constBits(),
+                    frame.sizeInBytes() / 2,
+                    screenDebug.bits());
+            break;
+        case LegacyControllerMapping::ScreenInfo::ColorEndian::Little:
+            qFromLittleEndian<ushort>(frame.constBits(),
+                    frame.sizeInBytes() / 2,
+                    screenDebug.bits());
+            break;
+        default:
+            break;
+        }
+        if (screeninfo.reversedColor) {
+            screenDebug.rgbSwap();
+        }
+
+        emit previewRenderedScreen(screeninfo, screenDebug);
+    }
+
+    QByteArray input((const char*)frame.constBits(), frame.sizeInBytes());
```

Comment:
would a static cast also work? c-style casts can be nasty.

Reply:
I'm getting `error: static_cast from 'const uchar *' (aka 'const unsigned char *') to 'const char *' is not allowed`. Not quite sure why it wouldn't be allowed... Any suggestion?

Reply:
Looks like `reinterpret_cast` did the trick

Reply:
Yeah. Still not ideal technically.

> The mapping performed by reinterpret_cast is implementation-defined. [Note: it might, or might not, produce a representation different from the original value. ]" (5.2.10/3)

[std::bit_cast](https://en.cppreference.com/w/cpp/numeric/bit_cast) seems good then.